File: ovinger/oving_1/shared variable/c/foo2.c
// Compile with `gcc foo.c -Wall -std=gnu99 -lpthread`, or use the makefile
// The executable will be named `foo` if you use the makefile, or `a.out` if you use gcc directly

// test med mutexes. 

#include <pthread.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>

int i = 0;
pthread_t thread1;
pthread_t thread2;
pthread_mutex_t mutex;

// Note the return type: void*
void* incrementingThreadFunction(){
    // TODO: increment i 1_000_000 timesmak
    for (int j = 0; j < 10; j++) {
        pthread_mutex_lock(&mutex);
        i++;
        printf("Incrementing: %d\n", i);
        pthread_mutex_unlock(&mutex);
        sleep(1);
    }
    return NULL;
}

void* decrementingThreadFunction(){
    // TODO: decrement i 1_000_000 times
    for (int j = 0; j < 9; j++) {
        pthread_mutex_lock(&mutex);
        i--;
        printf("Decrementing: %d\n", i);
        pthread_mutex_unlock(&mutex);
        sleep(1);
    }
    return NULL;
}


int main(){
    // TODO: 
    // start the two functions as their own threads using `pthread_create`
    // Hint: search the web! Maybe try "pthread_create example"?
    pthread_mutex_init(&mutex, NULL);

    int error1 = pthread_create(&thread1, NULL, *incrementingThreadFunction, &i);
    int error2 = pthread_create(&thread2, NULL, *decrementingThreadFunction, &i);

    error1  = 2;
    if (error1 || error2) {
        printf("Error creating threads\n Error1: %s\n Error2: %s\n", strerror(error1), strerror(error2));
        //return 1;
    }

    pthread_join(thread1, NULL);
    pthread_join(thread2, NULL);
    
    pthread_mutex_destroy(&mutex);
    printf("The magic number is: %d\n", i);
    return 0;
}

File: ovinger/oving_1/shared variable/go/foo2.go
// Use `go run foo.go` to run your program

package main

import (
    . "fmt"
    "runtime"
)

var i = 0

func incrementing(done chan bool) {
    //TODO: increment i 1000000 times
    for j := 0; j < 1000000; j++ {
        i++
    }
	done <- true
}

func decrementing(done chan bool) {
    //TODO: decrement i 1000000 times
    for j := 0; j < 1000000; j++ {
        i--
    }
	done <- true
}

func main() {
    // What does GOMAXPROCS do? What happens if you set it to 1?
    runtime.GOMAXPROCS(2)    
	// two threads to run the two functions concurrently

	done := make(chan bool, 2)

    // TODO: Spawn both functions as goroutines
    go incrementing(done)
    go decrementing(done)

	for k:= 0; k < 2; k++{
		select {
			case <- done:
		}
	}
    // We have no direct way to wait for the completion of a goroutine (without additional synchronization of some sort)
    // We will do it properly with channels soon. For now: Sleep.
    Println("The magic number is:", i)
}
